### ISSUE TITLE
AccessToken valid duration and RefreshToken usage

### ISSUE THREAD

**Issue by franky1017** [BitzArt/Blazor.Auth#17]
Been using your component and enjoyed it. Quick question - will your component automatically and periodically call the RefreshJwtPairAsync() function? Is the purpose to retrieve a new AccessToken? If so, what is the automatic period? How is that related to the validation period of the AccessToken?

Also, if the purpose of the RefreshJwtPairAsync() is to retrieve a new AccessToken when it expires, does it mean user will never need to re-login?

Thanks!

Frank

**Comment by YuriyDurov** [comment]
@franky1017 

`RefreshJwtPairAsync` method of your `AuthenticationStateProvider` will be called in case if `GetAuthenticationStateAsync` method of `AuthenticationStateProvider` was called after access token has expired (and refresh token has not).

Access token and refresh token expire automatically based on expiration timestamps that you can provide in the `JwtPair`.

`GetAuthenticationStateAsync` is being called by Blazor on every page pre-render, and also once per [scope](https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/dependency-injection) in interactive render modes.

You can also manually inject `AuthenticationStateProvider` anywhere in your app and call it's `GetAuthenticationStateAsync` method - this will trigger the refresh if the access token has expired.

User will never need to re-login if they always refresh in time (meaning while they have a refresh token that is not expired), assuming you allow them to refresh indefinitely. Normally, you would probably want to implement a limit on how long a chain of refreshes can be until the user needs to re-login, or restricting the refresh after some period of time from the original token's time of issue has passed.

**Comment by luis-fss** [comment]
Hi, @YuriyDurov, first of all I wanted to say that Blazor.Auth solved almost all my authentication and authorization problems in Blazor, thank you very much for that.
May I ask... when a refresh token expires during a form post, is there any way to redirect to the login page and then return to the previous page?
Or better yet, would it be possible to revalidate the login state (for example using a modal window) without requiring a full refresh on the current page, that is, without losing the data entered in the form?

**Comment by YuriyDurov** [comment]
@luis-fss 

It's great that the package is useful!

Do you really mean "when the refresh token expires" or actually it is "when the access token expires"?

Do you find it a common occurrence that your refresh token expires? What is the duration of your refresh token? It should be a somewhat rare event.

**Comment by luis-fss** [comment]
I refer to the refresh token.
I'll try to describe my case: my access token lasts 60 minutes and my refresh token lasts 7 days. So, from what I understand, every hour the access token will be renewed until the end of the seventh day, which is when the refresh token will have expired and a new login will be necessary, correct?

The problem here is that in almost all internal operations that require storing data in the database I also need to know the user's name. I would like to know if when I execute the "GetAuthenticationStateAsync" command and the return is null or "not authenticated" would it be possible to revalidate the login state without losing the state of the current component/page?
What I do now is simply present a message to the user that they will need to log in again, but this will also cause them to lose all the information entered into a form.

But you're right, it's a rare event, although it could still happen once a week with my setup. And it doesn't have many negative effects on the user experience, so don't worry about it if it's an edge case or if it's complex to troubleshoot.
As I said, Blazor.Auth already helps me a lot and I think it should be Blazor's default authentication solution.

**Comment by YuriyDurov** [comment]
@luis-fss on every jwt pair refresh (every hour, in your case), the refresh token itself should also refresh. Just issue a new refresh token whenever the access token expires. In this case, the refresh token can only expire if the user did not come online for the duration of your refresh token (a week, in your case).

**Blazor.Auth** works assuming you will provide a new jwt pair on every refresh.

**Comment by luis-fss** [comment]
@YuriyDurov oh, so I was misunderstanding.
Thanks for clarifying.

**Comment by YuriyDurov** [comment]
@luis-fss **Blazor.Auth** does not implement the actual logic of generating the JWT pair but leaves it to the developer - this way it can be adapted to your specific needs. No end implementation is wrong, I am only suggesting a flow that I see as optimal for a somewhat standard solution. Refresh it or not - in the end, it's up to you.

**Comment by luis-fss** [comment]
@YuriyDurov yes, "No end implementation is wrong" I get that, thanks.
I was renewing the refresh token only during the login process, but you gave me the idea of ​​renewing it together with the access token, which is perhaps the most common and industry standard way or perhaps not, but I think it will solve the problem in my case. Thanks a lot.

**Comment by franky1017** [comment]
@YuriyDurov 

I wanted to see this in action, so I modified my access token to expires in 10 seconds and refresh token like a day. I put a debug break point on the function RefreshJwtPairAsync(). After login, I continue to click on different pages to be active and after 10 seconds, I got an api error about invalid user (as the access token has expired).

The RefreshJwtPairAsync() function was never called during my debug session. And the 10 seconds access token just expired on its own.

Did I miss any configuration settings? Based on the discussion in this thread, the access token should be automatically refreshed by calling the RefreshJwtPairAsync() function every 10 seconds?

Frank
